src/Controller/AppController.php

```diff
@@ -124,7 +124,13 @@ public function beforeFilter(Event $event)
 
         // So that we can access the current users info from models.
         // Important: needs to be done after RememberMe->check().
-        CurrentUser::store($this->Auth->user());
+        $user = $this->Auth->user();
+        if ($user) {
+            // Keep the info up to date
+            $this->loadModel('Users');
+            $user = $this->Users->getInformationOfCurrentUser($user['id'])->toArray();
+        }
+        CurrentUser::store($user);
```

Comment:
While this definitely fixes this issue (as well as the issue of having to logout and login after a user status change), I am concerned with the design of pooling the users table on each and every request, most of the time needlessly. Pooling is also what we do to display the unread private message counter at the top-right, and it bloats the list of fixtures required to run most of our controller tests. I feel like it’s not a good design and I wonder if there is any better way to go.

As for the profile picture issue, you could as well update the session data after saving it. That’s what we do everywhere else, for example:
https://github.com/Tatoeba/tatoeba2/blob/9850fd42d3d70f6472c77e16668ecb4a7775e176/src/Controller/UserController.php#L318-L325

Reply:
Well I’ve thought it over and I can’t think of a better way to go. The current implementation fails in a lot of ways, including when a user logs in from two different browsers simultaneously. We already pool the users_languages table in `CurrentUser::_setProfileLanguages()` anyway. You could then remove the call to `setUser()` that I mentioned, as well as other similar calls.

My only remaining concern is that I think (haven’t tested) that the Auth component won’t know about a user role update because you don’t run `$this->Auth->setUser()` with the latest user data. Because of this, the restrictions defined in `config/auth_actions.php` won’t apply match to the new role.

Reply:
Aside: I think you mean "poll" instead of "pool".

> You could then remove the call to `setUser()` that I mentioned, as well as other similar calls.

Will do.

> the Auth component won’t know about a user role update because you don’t run `$this->Auth->setUser()` with the latest user data.

I was actually calling `$this->Auth->setUser()` at first, but then the request failed with an "auth error" (if I remember correctly), so I figured that the `Auth` component isn't supposed to be used like that.

> Because of this, the restrictions defined in `config/auth_actions.php` won’t apply match to the new role.

Thanks for pointing these out to me. I thought the permission checks all went through `CurrentUser::canEditSentenceOfUserId` and the like, so it would be enough to keep `CurrentUser` up-to-date. I'll look into updating the `Auth` user data as well.

Reply:
> So I opted to update `CurrentUser` after calls to `Users->save`.

Scratch that part. I tested the wrong commit. Turns out `beforeRender` doesn't get called if there's nothing to render because the request is redirected. So `Auth->setUser` has to be called directly to make changes take effect immediately.

Reply:
> I thought that would make the other `setUser` calls entirely redundant, but one of the tests didn't agree. When changing the user data, there would a delay of one request until the new data reaches `Auth->user`. So I opted to update `CurrentUser` after calls to `Users->save`.

Hah, I bet you’re talking about `testSaveBasic_changingEmailUpdatesAuthData()`! I wrote it long ago because of a similar bug #526. Of course the test fails because it only performs one request; it’s tied to that design of calling `Auth->setUser()` on user "write" rather than "read". Fell free to rewrite or remove that test.

Reply:
I've rewritten the test to actually follow the redirect before checking that the Auth data was changed.